Common traits PartialEq, Eq and Hash

dns2utf8 opened the pull request:
I implemented PartialEq, Eq and Hash. Others like Copy or PartialOrd did not seem right.

dns2utf8 commented:
Currently the build fails prior to rustc 1.12.0 because of the `assert_ne!` macro.
The `DefaultHasher` requires rustc 1.13.0. 

Should we work around it or update the min version? Debian 9 stable currently serves rust 1.14.0 with the default installation.

frewsxcv commented:
`PartialEq` and `Eq` make a bit of sense to me. Not sure why you'd need to compare threadpools in the real world, but sure, I think this is fine. I'm a little confused by `Hash` though. Are you aware of any real world examples involving `Hash`? Just trying to better understand why the trait implementation is being added

dns2utf8 commented:
My intention was to implement everything I thought reasonable.
I made a test case with the hash set. I agree do not see an imminent real-world use-case for the hash trait yet.

I implemented it from the list of [this document](https://public.etherpad-mozilla.org/p/rust-api-guidelines-threadpool) referenced [here](https://internals.rust-lang.org/t/out-of-band-crate-evaluation-for-2017-07-06-threadpool/5495):
Types eagerly implement common traits ([C-COMMON-TRAITS])
    - `Copy`, `Clone`, `Eq`, `PartialEq`, `Ord`, `PartialOrd`, `Hash`, `Debug`, `Display`, `Default`

We could also split this PR and only merge the `PartialEq` and `Eq`.

dns2utf8 commented:
Cool, thank you!
